### PR TITLE
Find optimization

### PR DESCRIPTION
**Pull request by snarkipus**
Added test case - hopefully I did this correctly.

### CODE REVIEW

File: packages/langium/src/utils/cst-util.ts [hunk 1]
```diff
@@ -109,12 +109,19 @@ export function findLeafNodeAtOffset(node: CstNode, offset: number): LeafCstNode
...
+            let middleChild = Math.floor((firstChild + lastChild) / 2);
+            const n = node.children[middleChild];
+            if (n.offset > offset) {
+                lastChild = middleChild - 1;
+            } else if (n.end <= offset) {
+                firstChild = middleChild + 1;
+            } else {
+                return findLeafNodeAtOffset(n, offset);
             }
+            middleChild = Math.floor((firstChild + lastChild) / 2);
```
Comment (spoenemann) [inline packages/langium/src/utils/cst-util.ts:124]: `middleChild` is already set at the beginning of the iteration, no need to set it again.

File: packages/langium/test/utils/cst-utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    test('Find Leaf Node at Offset', async () => {
+        const text = `
+        grammar test
+        Main: value=AB;
+        terminal fragment Frag: 'B';
+        terminal AB: 'A' Frag;
+        `;
+
+        const grammar = await parser(text);
+        const offset = grammar.parseResult.value.$document?.textDocument.getText().indexOf('AB');
```
Comment (spoenemann) [inline packages/langium/test/utils/cst-utils.test.ts:24]:
```suggestion
        const offset = grammar.textDocument.getText().indexOf('AB');
```

**Review by spoenemann — COMMENTED** [review]
Thanks, almost good to go.

File: packages/langium/test/utils/cst-utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        const leafnode = c.findLeafNodeAtOffset(grammar.parseResult.value.$cstNode!, offset!);
```
Comment (msujew) [inline packages/langium/test/utils/cst-utils.test.ts:25]:
```suggestion
        const leafnode = c.findLeafNodeAtOffset(grammar.$cstNode!, offset!);
```

File: packages/langium/test/utils/cst-utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+import { createLangiumGrammarServices, EmptyFileSystem, Grammar } from '../../src';
+import { parseHelper } from '../../src/test';
+import * as c from '../../src/utils/cst-util';
```
Comment (msujew) [inline packages/langium/test/utils/cst-utils.test.ts:8]: Minor: The name is a bit non-descriptive.
```suggestion
import * as cstUtil from '../../src/utils/cst-util';
```

File: packages/langium/test/utils/cst-utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+const services = createLangiumGrammarServices(EmptyFileSystem);
+const parser = parseHelper<Grammar>(services.grammar);
+
+describe('CST Utils', () => {
+
+    test('Find Leaf Node at Offset', async () => {
```
Comment (msujew) [inline packages/langium/test/utils/cst-utils.test.ts:15]: Suggestion: It might be interesting to test edge case behavior in here as well. Your test covers the case where the cursor is at `<|>AB;`, but `A<|>B` and `AB<|>;` (which should return the leaf node for `;`) are also very important.

File: packages/langium/test/utils/cst-utils.test.ts (hunk 2 above)
Comment (snarkipus) [reply packages/langium/test/utils/cst-utils.test.ts]: fixed

File: packages/langium/test/utils/cst-utils.test.ts (hunk 3 above)
Comment (snarkipus) [reply packages/langium/test/utils/cst-utils.test.ts]: From what I could tell, `$cstNode` is only available on the `parseResult`. Am I missing something?

File: packages/langium/test/utils/cst-utils.test.ts (hunk 4 above)
Comment (snarkipus) [reply packages/langium/test/utils/cst-utils.test.ts]: fixed

File: packages/langium/test/utils/cst-utils.test.ts (hunk 5 above)
Comment (snarkipus) [reply packages/langium/test/utils/cst-utils.test.ts]: added add'l test cases

File: packages/langium/test/utils/cst-utils.test.ts (hunk 3 above)
Comment (msujew) [reply packages/langium/test/utils/cst-utils.test.ts]: Yeah, sorry got confused, since I expected `grammar` to be an AST element, not a document. It's all good.

**Review by spoenemann — APPROVED** [review]

**Comment by spoenemann** [comment]
Unrelated to this change: we don't check the offset if the given node is already a leaf node. @msujew could there be cases where this is problematic?
